src/Interpreters/MonotonicityCheckVisitor.h

```diff
@@ -0,0 +1,138 @@
+#pragma once
+
+#include <AggregateFunctions/AggregateFunctionFactory.h>
+#include <DataTypes/DataTypeFactory.h>
+#include <Functions/FunctionFactory.h>
+#include <IO/WriteHelpers.h>
+#include <Interpreters/InDepthNodeVisitor.h>
+#include <Interpreters/IdentifierSemantic.h>
+#include <Parsers/ASTFunction.h>
+#include <Parsers/ASTIdentifier.h>
+#include <Parsers/ASTOrderByElement.h>
+#include <Parsers/ASTSelectQuery.h>
+#include <Parsers/ASTTablesInSelectQuery.h>
+#include <Parsers/IAST.h>
+#include <Common/typeid_cast.h>
+
+namespace DB
+{
+
+using Monotonicity = IFunctionBase::Monotonicity;
+
+/// Checks from bottom to top if function composition is monotonous
+class MonotonicityCheckMatcher
+{
+public:
+    struct Data
+    {
+        const TablesWithColumns & tables;
+        const Context & context;
+        const std::unordered_map<String, ASTPtr> & group_by_function_hashes;
+        Monotonicity monotonicity{true, true, true};
+        ASTIdentifier * identifier = nullptr;
+        DataTypePtr arg_data_type = {};
+
+        void reject() { monotonicity.is_monotonic = false; }
+        bool isRejected() const { return !monotonicity.is_monotonic; }
+
+        bool canOptimize(const ASTFunction & ast_function) const
+        {
+            /// if GROUP BY contains the same function ORDER BY shouldn't be optimized
+            auto hash = ast_function.getTreeHash();
+            String key = toString(hash.first) + '_' + toString(hash.second);
+            if (group_by_function_hashes.count(key))
+                return false;
+
+            /// if ORDER BY contains aggregate function it shouldn't be optimized
+            if (AggregateFunctionFactory::instance().isAggregateFunctionName(ast_function.name))
+                return false;
+
+            return true;
+        }
+
+        bool getIdentifierAndType(const ASTFunction & ast_function)
+        {
+            identifier = ast_function.arguments->children[0]->as<ASTIdentifier>();
```

Comment:
It's not clear how it works for nested functions. Maybe add a comment, that we call this function only for the most bottom argument (I can see such comment bellow).

Reply:
Noone believes such comments :)
I've add precondition about `identifier` that makes the function more self descriptive.